### PR TITLE
Implement extraction of org memory quota metrics

### PR DESCRIPTION
**Pull request by bkabrda**
*Note: Please remember to review the Datadog [Contribution Guidelines](https://github.com/DataDog/dd-agent/blob/master/CONTRIBUTING.md)
if you have not yet done so.*

### What does this PR do?

Implements extraction of Org memory quotas for all orgs. There's one TODO left in the code - I'm uncertain whether we want to provide a `host` for the org quota metrics. I think not, as orgs are not tied to a specific host, but I'd like to hear more opinions.

### Motivation

What inspired you to submit this pull request?

### Testing Guidelines

An overview on [testing](https://github.com/DataDog/dd-agent/blob/master/tests/README.md)
is available in our contribution guidelines.

### Additional Notes

Anything else we should know when reviewing?

### CODE REVIEW

File: internal/orgcollector/orgcollector.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+	wg.Add(1)
+	var allOrgs []cfclient.Org
+	go func() {
+		defer wg.Done()
+		var err error
+		allOrgs, err = o.cfClient.GetV2Orgs()
+		if err != nil {
+			errors <- err
+		}
+	}()
```
Comment (zippolyte) [inline internal/orgcollector/orgcollector.go]: Can't we use the org information that is already cached, instead of fetching it every time ?

File: internal/orgcollector/orgcollector.go [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+		allQuotas, err = o.cfClient.GetV2OrgQuotas()
+		if err != nil {
+			errors <- err
+		}
+	}()
+
+	wg.Wait()
+	close(errors)
+
+	// Go through the channel, print all errors and return one of them if there are any
```
Comment (zippolyte) [inline internal/orgcollector/orgcollector.go]:
```suggestion
	// Go through the channel and print all errors
```

File: internal/orgcollector/orgcollector.go [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+		tags := []string{}
+		tags = append(tags, o.customTags...)
+		tags = append(tags, o.getTagsFromOrg(org)...)
+		key := metric.MetricKey{
+			Name:     "org.memory.quota",
+			TagsHash: util.HashTags(tags),
+		}
+		value := metric.MetricValue{
+			Tags: tags,
+			// TODO: what host do we want to provide?
```
Comment (zippolyte) [inline internal/orgcollector/orgcollector.go]: I don't think we need any host

File: internal/orgcollector/orgcollector.go [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+	}
+	o.processedMetrics <- metricsPackages
+}
+
+func (o *OrgCollector) getTagsFromOrg(org cfclient.Org) []string {
+	tags := []string{}
+	tags = append(tags, fmt.Sprintf("guid:%s", org.Guid))
+	tags = append(tags, fmt.Sprintf("org_name:%s", org.Name))
+	tags = append(tags, fmt.Sprintf("org_id:%s", org.Guid))
+	tags = append(tags, fmt.Sprintf("status:%s", org.Status))
```
Comment (zippolyte) [inline internal/orgcollector/orgcollector.go]: What's the status of an org ?

File: internal/client/cloudfoundry/client.go [hunk 5]
```diff
@@ -515,6 +515,48 @@ func (cfc *CFClient) getV2ApplicationsByPage(page int) ([]CFApplication, int, er
...
+	query.Set("results-per-page", "100")
+
+	for p := 1; ; p++{
+		query.Set("page", strconv.Itoa(p))
+		moreOrgs, err := cfc.client.ListOrgsByQuery(query)
+		if err != nil {
+			return nil, err
+		}
+		allOrgs = append(allOrgs, moreOrgs...)
+		if len(moreOrgs) < 100 {
```
[the remark this answers is not in the document]
Comment (bkabrda) [reply internal/client/cloudfoundry/client.go]: Heh, this was a good question. It led me to discover that we don't actually need to do the loop, as the cfclient does it itself. So I'll just drop the loop completely (and FTR if there's no org on page 1, the function will return an all-ok response with empty list of orgs).

File: internal/orgcollector/orgcollector.go [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+	"github.com/cloudfoundry-community/go-cfclient"
+	"github.com/cloudfoundry/gosteno"
+)
+
+type OrgCollector struct {
+	cfClient              *cloudfoundry.CFClient
+	log                   *gosteno.Logger
+	processedMetrics      chan<- []metric.MetricPackage
+	customTags            []string
+	queryInterval		  uint32
```
[the remark this answers is not in the document]
Comment (bkabrda) [reply internal/orgcollector/orgcollector.go]: Thanks, will fix.

File: internal/orgcollector/orgcollector.go (hunk 1 above)
Comment (bkabrda) [reply internal/orgcollector/orgcollector.go]: I wanted to do that, but IMO it's not worth the effort right now. Let me explain: We either get app cache from V3 entirely or V2 entirely.

* If V3 is available (which is true most often on currently supported CF versions), we get everything from V3
  * V3 only has experimental support for returning quota GUID as a member of the org structure and it's not even present in all supported CF versions, so we can't rely on that anyway. Quota definitions are not available through V3 in *any* CF version yet.
  * This means V3 is out of question.
* If only V2 is available, we query only the app endpoint and make it return related values (including orgs). So we'd have to special case operations with V2 to also store the Orgs, which wouldn't be completely straightforward and it would bring little value (IMO), since we all currently supported CF versions should already have V3 and we can't use that for orgs as explained above.

Basically I think we should do this at the point when all supported CF versions have quotas available through the V3 API (which will probably take a while) - I also made a comment about this at the top of the `orgcollector.go`.

File: internal/orgcollector/orgcollector.go (hunk 4 above)
Comment (bkabrda) [reply internal/orgcollector/orgcollector.go]: It's either `active` or `suspended` [1]. I think it may be useful for admins to e.g. only see (or sum up) quotas for active orgs.

[1] https://docs.cloudfoundry.org/concepts/roles.html#orgs

File: internal/orgcollector/orgcollector.go (hunk 3 above)
Comment (bkabrda) [reply internal/orgcollector/orgcollector.go]: 👍

**Comment by bkabrda** [comment]
@zippolyte thanks for your review! I think I've addressed all the points, please take a look.

File: internal/orgcollector/orgcollector.go (hunk 1 above)
Comment (zippolyte) [reply internal/orgcollector/orgcollector.go]: OK fair enough. Thanks for the explanation. Do we have an idea of the cardinality of orgs in the bigger infrastructures ?

File: internal/orgcollector/orgcollector.go (hunk 4 above)
Comment (zippolyte) [reply internal/orgcollector/orgcollector.go]: 👍

File: internal/client/cloudfoundry/client.go [hunk 7]
```diff
@@ -517,41 +517,25 @@ func (cfc *CFClient) getV2ApplicationsByPage(page int) ([]CFApplication, int, er
 
 func (cfc *CFClient) GetV2Orgs() ([]cfclient.Org, error) {
 	// NOTE: this is used by the OrgCollector
-	var allOrgs []cfclient.Org
 	query := url.Values{}
-	query.Set("results-per-page", "100")
+	query.Set("results-per-page", "1")
```
Comment (zippolyte) [inline internal/client/cloudfoundry/client.go]: Is this on purpose ?

Comment (bkabrda) [reply internal/client/cloudfoundry/client.go]: Whoops, good catch :) I did some testing here... sorry, will fix :)

Comment (bkabrda) [reply internal/client/cloudfoundry/client.go]: Fixed.

File: internal/orgcollector/orgcollector_test.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+		fakeOrgCollector, err = NewOrgCollector(
+			&cfg,
+			pm,
+			log,
+			customTags,
+		)
+		Expect(err).To(BeNil())
+	}, 0)
+
+	It("pushes correct metrics", func() {
```
Comment (gzussa) [inline internal/orgcollector/orgcollector_test.go]: 👍

File: test/helper/fake_cloud_controller_api.go [hunk 9]
```diff
@@ -2823,6 +2834,118 @@ func (f *FakeCloudControllerAPI) writeResponse(rw http.ResponseWriter, r *http.R
 					]
 				}`)))
 		}
+	case "/v2/quota_definitions":
+		rw.Write([]byte(fmt.Sprintf(`{
```
Comment (gzussa) [inline test/helper/fake_cloud_controller_api.go]: Something i wanted to do here (for clean up) was to put all those JSON in resource files and load them here instead.

File: internal/client/cloudfoundry/client.go [hunk 10]
```diff
@@ -515,6 +515,32 @@ func (cfc *CFClient) getV2ApplicationsByPage(page int) ([]CFApplication, int, er
 	return results, appResp.Pages, nil
 }
 
+func (cfc *CFClient) GetV2Orgs() ([]cfclient.Org, error) {
+	// NOTE: this is used by the OrgCollector
```
Comment (gzussa) [inline internal/client/cloudfoundry/client.go]: I think we can remove these notes. The compiler would tell you if you dare to remove these methods 😛

File: internal/orgcollector/orgcollector.go [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+	cfClient, err := cloudfoundry.NewClient(config, log)
+	if err != nil {
+		return nil, err
+	}
+	return &OrgCollector{
+		cfClient:         cfClient,
+		log:              log,
+		processedMetrics: processedMetrics,
+		customTags:       customTags,
+		queryInterval:    config.OrgDataQuerySeconds,
```
Comment (gzussa) [inline internal/orgcollector/orgcollector.go]: This config name is confusing to me. Shouldn't it be better to call it `OrgDataCollectionInterval` or something that makes it obvious it is an interval?

File: internal/orgcollector/orgcollector.go [hunk 12]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+func (o *OrgCollector) pushMetrics() {
+	o.log.Info("Collecting org quotas ...")
+	var wg sync.WaitGroup
+	errors := make(chan error, 10)
+	// Fetch orgs
+	wg.Add(1)
+	var allOrgs []cfclient.Org
+	go func() {
```
Comment (gzussa) [inline internal/orgcollector/orgcollector.go]: Do we need this go routine since the whole thing is already a go routine?

Comment (gzussa) [reply internal/orgcollector/orgcollector.go]: Oh! 🤔 Is it so that we call both `GetV2Orgs` and `GetV2OrgQuotas` in parallel? Statistically speaking I wonder if it is worse it since I doubt that we would have more than 100 orgs in general. What do you think, should we simplify this code in order to gain clarity?

File: internal/orgcollector/orgcollector.go (hunk 1 above)
Comment (bkabrda) [reply internal/orgcollector/orgcollector.go]: From what I've been told it's in high hundreds, maybe a thousand or a little bit more.

File: internal/orgcollector/orgcollector.go (hunk 3 above)
Comment (gzussa) [reply internal/orgcollector/orgcollector.go]: Let's remove this TODO

File: config/datadog-firehose-nozzle.json [hunk 13]
```diff
@@ -22,5 +22,6 @@
   "CloudControllerAPIBatchSize": 500,
   "AppMetrics": true,
   "NumWorkers": 1,
-  "CustomTags": []
+  "CustomTags": [],
+  "OrgDataQuerySeconds": 600
```
Comment (gzussa) [inline config/datadog-firehose-nozzle.json]: Good call to make this configurable 👏

File: test/helper/fake_cloud_controller_api.go (hunk 9 above)
Comment (bkabrda) [reply test/helper/fake_cloud_controller_api.go]: Agreed, but let's do that in a separate PR.

File: internal/client/cloudfoundry/client.go (hunk 10 above)
Comment (bkabrda) [reply internal/client/cloudfoundry/client.go]: True, I'll remove them :)

File: internal/orgcollector/orgcollector.go (hunk 11 above)
Comment (bkabrda) [reply internal/orgcollector/orgcollector.go]: You're right, that does sound better. I'll change it.

File: internal/orgcollector/orgcollector.go (hunk 12 above)
Comment (bkabrda) [reply internal/orgcollector/orgcollector.go]: I hear that some users have around 1000 orgs or even more, so I think it's worth parallelizing (there will necessarily be at least the same amount of quotas).

**Comment by bkabrda** [comment]
Thanks for all the reviews! Merging.
